### PR TITLE
add function instance injector supports for dependency injection

### LINKED ISSUE #310: Hooks for DI frameworks
Function classes are instantiated directly via Class.newInstance().  Would be great if Functions has some simple hooks to integrate DI frameworks (Spring, Guice, Arc, CDI).  A function deployment could publish a simple interface:

```
interface FunctionFactory {
    <T> T newInstance(Class<T> functionClass);
}
```

This interface could be discovered via a [ServiceLoader ](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html) bundled in the app jar, and/or through function.json config.

FunctionFactory should maybe be within azure-functions-java-library?

### PR DESCRIPTION
**Pull request by kaibocai**
### Issue describing the changes in this PR

resolves [#issue_for_this_pr](https://github.com/Azure/azure-functions-java-worker/issues/310)

### Pull request checklist

* [ ] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 1]
```diff
@@ -44,26 +46,57 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+				WorkerLogManager.getSystemLogger().info("Didn't find any function instance injector, creating class instance every invocation.");
+			}
+			if (iterator.hasNext()){
+				WorkerLogManager.getSystemLogger().warning("Customer function app has multiple FunctionInstanceInjector implementations.");
+				throw new RuntimeException("Customer function app has multiple FunctionInstanceInjector implementations");
 			}
-			middlewares.add(getFunctionExecutionMiddleWare());
-			this.invocationChainFactory = new InvocationChainFactory(middlewares);
+		} finally {
+			Thread.currentThread().setContextClassLoader(ClassLoader.getSystemClassLoader());
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]:
```suggestion
			Thread.currentThread().setContextClassLoader(prevContextClassLoader);
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 2]
```diff
@@ -44,26 +46,57 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+					public <T> T getInstance(Class<T> functionClass) throws Exception {
+						return functionClass.newInstance();
+					}
+				};
+				WorkerLogManager.getSystemLogger().info("Didn't find any function instance injector, creating class instance every invocation.");
+			}
+			if (iterator.hasNext()){
+				WorkerLogManager.getSystemLogger().warning("Customer function app has multiple FunctionInstanceInjector implementations.");
+				throw new RuntimeException("Customer function app has multiple FunctionInstanceInjector implementations");
 			}
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]: can move up into if condition

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 3]
```diff
@@ -44,26 +46,57 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+				this.functionInstanceInjector = iterator.next();
+				WorkerLogManager.getSystemLogger().info("Load function instance injector: " + this.functionInstanceInjector.getClass().getName());
+			}else {
+				this.functionInstanceInjector = new FunctionInstanceInjector() {
+					@Override
+					public <T> T getInstance(Class<T> functionClass) throws Exception {
+						return functionClass.newInstance();
+					}
+				};
+				WorkerLogManager.getSystemLogger().info("Didn't find any function instance injector, creating class instance every invocation.");
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]:
```suggestion
				this.functionInstanceInjector = JavaFunctionBroker::newInstance;
				WorkerLogManager.getSystemLogger().info("Didn't find any function instance injector, creating class instance every invocation.");
```

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 4]
```diff
@@ -44,26 +46,57 @@ public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingI
...
+	}
+
+	private void initializeFunctionInstanceInjector() {
+		try {
+			//ServiceLoader will use thread context classloader to verify loaded class
+			Thread.currentThread().setContextClassLoader(classLoaderProvider.createClassLoader());
+			Iterator<FunctionInstanceInjector> iterator = ServiceLoader.load(FunctionInstanceInjector.class).iterator();
+			if (iterator.hasNext()) {
+				this.functionInstanceInjector = iterator.next();
+				WorkerLogManager.getSystemLogger().info("Load function instance injector: " + this.functionInstanceInjector.getClass().getName());
```
Comment (trask) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java:97]:
```suggestion
				WorkerLogManager.getSystemLogger().log(Level.INFO, "Load function instance injector: {0}", this.functionInstanceInjector.getClass().getName());
```

Comment (kaibocai) [reply src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java:97]: Hi @trask , I only found two `info` methods in `java.utli.logging`, [one ](https://docs.oracle.com/javase/8/docs/api/java/util/logging/Logger.html#info-java.lang.String-)takes one `String` parameter, the [other one](https://docs.oracle.com/javase/8/docs/api/java/util/logging/Logger.html#info-java.util.function.Supplier-) takes one `Supplier<String>` as parameter. Seems `java.utli.logging` doesn't support formatting log like above.

**Review by trask — DISMISSED** [review]

File: src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java [hunk 5]
```diff
@@ -93,7 +94,7 @@ private void initializeFunctionInstanceInjector() {
 			Iterator<FunctionInstanceInjector> iterator = ServiceLoader.load(FunctionInstanceInjector.class).iterator();
 			if (iterator.hasNext()) {
 				this.functionInstanceInjector = iterator.next();
-				WorkerLogManager.getSystemLogger().info("Load function instance injector: " + this.functionInstanceInjector.getClass().getName());
+				WorkerLogManager.getSystemLogger().log(Level.INFO, "Load function instance injector: {0}" + this.functionInstanceInjector.getClass().getName());
```
Comment (kaibocai) [inline src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java]: some how this logs something like "Load function instance injector: {0}" in kusto, not log out the class name. I am thinking to switch to using String.format, once we figure out why it not works and fix it, we can have it change back into this way.

**Review by trask — DISMISSED** [review]

**Pull request merged**